workers/gossip-server/src/peer_discovery/expiry_window.rs

```diff
@@ -0,0 +1,169 @@
+//! Defines a windowed buffer that manages the state transitions of peers in the
+//! network
+
+use std::{
+    collections::HashMap,
+    hash::Hash,
+    time::{Duration, Instant},
+};
+
+use common::{new_async_shared, types::gossip::WrappedPeerId, AsyncShared};
+use tokio::sync::{RwLockReadGuard, RwLockWriteGuard};
+
+/// The amount of time other cluster peers are allowed to give liveness
+/// attestations for an expiry candidate
+pub(crate) const EXPIRY_CANDIDATE_WINDOW_MS: u64 = 10_000; // 10 seconds
+/// The minimum amount of time between a peer's expiry and when it can be
+/// added back to the peer info
+pub(crate) const EXPIRY_INVISIBILITY_WINDOW_MS: u64 = 30_000; // 30 seconds
+
+/// A buffer of time windows
+#[derive(Clone)]
+pub struct TimeWindowBuffer<T: Hash + Eq + PartialEq> {
+    /// The set of time windows; maps a key to the time at which it should exit
+    /// the buffer
+    ///
+    /// Thread safe
+    pub windows: AsyncShared<HashMap<T, Instant>>,
+}
+
+impl<T: Hash + Eq + PartialEq> TimeWindowBuffer<T> {
+    /// Constructor
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self { windows: new_async_shared(HashMap::new()) }
+    }
+
+    /// Acquire a read lock on the windows
+    async fn read_windows(&self) -> RwLockReadGuard<'_, HashMap<T, Instant>> {
+        self.windows.read().await
+    }
+
+    /// Acquire a write lock on the windows
+    async fn write_windows(&self) -> RwLockWriteGuard<'_, HashMap<T, Instant>> {
+        self.windows.write().await
+    }
+
+    /// Cleanup all expired windows
+    async fn cleanup(&self) {
+        let mut this = self.write_windows().await;
+        this.retain(|_, expiry_time| *expiry_time > Instant::now());
+    }
+
+    /// Whether or not the given key is present in the buffer
+    pub async fn contains(&self, key: &T) -> bool {
+        let this = self.read_windows().await;
+        this.contains_key(key)
+    }
+
+    /// Check whether the window for a key has expired
+    pub async fn is_expired(&self, key: &T) -> bool {
+        let this = self.read_windows().await;
+        match this.get(key) {
+            Some(expiry_time) => *expiry_time < Instant::now(),
+            None => true,
```

Comment:
thoughts on having this method return an `Option<bool>` and letting the callsite decide if a `None` implies the window has expired / was never added?

Reply:
For example if we call `PeerExpiryWindows::should_expire` below with a `peer_id` that was never a candidate, we will expire it